TT UI: style update

mmilata opened the pull request:
Fixes #2849, fixes #2734 and possibly #2657.

![4bf94745fa279464296871daeca7ae07aeb0327fbb9240f8fbf548e423cc4607](https://user-images.githubusercontent.com/85857/224579614-de677931-0cf6-47e0-991b-d48a7406e4a4.png) ![9af500194eb7a43f69a5c4bf13c9ec9a27ef32eea9fa637f532571be83f6dd49](https://user-images.githubusercontent.com/85857/224579615-d0f791bf-594c-4ecb-9813-7883df848cde.png) ![9b56f8f1d892732a729fc8508abb882efcb54d70e3c62fed3cb0f9ccf6c18dda](https://user-images.githubusercontent.com/85857/224579617-cdcfe196-b828-4370-9bf1-6f76bcf59719.png) ![474fd099177f48b1240ece11287e272a4ac463f0c6942a11c58a13ba76d796af](https://user-images.githubusercontent.com/85857/224579618-d54f9ec4-778a-4815-88fb-42b2f7e3fb36.png) ![9148aa6e5da50e3427f652d367aa1fe30fb5a65dec872fe4cedda7b45e427fe3](https://user-images.githubusercontent.com/85857/224579619-79de896a-4d64-477d-8d50-1ab02bd6a3bb.png) ![92919ac4ace757243d90aa35db64e7df48d49e7a327eb2d08beaf566fbd8fa29](https://user-images.githubusercontent.com/85857/224579620-046c593d-04b4-4bb1-a2f7-71ab3ccef602.png) ![f9d0f1f6d732fae0ee49cf3265c6f88a6140f2c42f828cbb362efbf834c5046f](https://user-images.githubusercontent.com/85857/224579621-0ae787d5-5035-4a5f-aebe-582048d9b3c2.png) ![f91c0f0165e1058670fd863e74c43343f5d3dcf2e214668d28610979c251179a](https://user-images.githubusercontent.com/85857/224579622-f25839ba-99cc-411e-8849-48173c97d803.png) ![fc7c6eb03347115fec4de5f255f16eb580d2b12f60aab3ec01506d0cd2289bb6](https://user-images.githubusercontent.com/85857/224579623-e7863064-8fe7-42dc-8399-dc169b08d666.png)

@TychoVrahe feel free to cherry-pick 5d556652a1c2cbe0ad2ffb426f6d1967129ec55d if you need the larger fonts for bootloader

Hannsek commented:
![image](https://user-images.githubusercontent.com/57008159/224471903-7f49aa3e-033c-4691-a640-68f1f5d226d1.png)
Please move the whole url to the new line

![image](https://user-images.githubusercontent.com/57008159/224472004-1955d86d-ee80-4bb5-9741-c99e3c622dc0.png)
Can we make the blur less blurry but darker? 
<img width="254" alt="image" src="https://user-images.githubusercontent.com/57008159/224472024-c7ed8432-ffa4-451b-a3ab-15fab240e0b3.png">

TychoVrahe commented:
The blurring radius can be changed, but this will affect user uploaded homescreens as well and from what i tested, less blurring was hardly noticable with some pictures - if you have image with sharp lines like this one it might be ok, but for everything else it probably isnt optimal.

Dimming requires recomputing every pixel, which might slow things here even further. But might be ok, needs to be tested.

Hannsek commented:
The problem here is that the “Locked, Tap to unlock” is not readable well enough. Making it only darker may solve this.

TychoVrahe commented:
I can see that, but it probably isnt worth increasing the rendering time to 300ms... not saying that it would, it just needs testing.

Addition is also computation. 

Or we could use some not so lame default background image :-)

mmilata commented:
This PR already increases `BLUR_SIZE` to 15, but I haven't tested it on the device ... will roll it back to 9 and see if making the screen darker instead works.

mmilata commented:
> will roll it back to 9 and see if making the screen darker instead works.

![92919ac4ace757243d90aa35db64e7df48d49e7a327eb2d08beaf566fbd8fa29](https://user-images.githubusercontent.com/85857/224578132-18c97ad1-9dac-4b29-b0a7-a9b5c76b05fd.png)

Close enough?

mmilata commented:
There's something wrong with `test_show_multisig_xpubs` but otherwise I think this is ready for review. [[UI diff](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/3918688814/artifacts/external_file/test_ui_report/differing_screens.html)][[differing screens](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3918688814/artifacts/test_ui_report/differing_screens.html)]

Hannsek commented:
The blur in now good. Thanks.

I noticed a differing spaces in the "locked" text.
![image](https://user-images.githubusercontent.com/57008159/224631107-3364b155-20f4-45ad-a391-9086f037cf51.png)
This is from the device. You can see big difference between the space between O and C and very small space between C and K.

This is from the figma: 
![image](https://user-images.githubusercontent.com/57008159/224631317-f6835fc8-213d-46fd-8227-8ffc4579cc36.png)

Hannsek commented:
On this screen 
![image](https://user-images.githubusercontent.com/57008159/224632042-0f8c8096-ec55-49e4-919e-e6ed24e213cf.png)
there are two inaccuracies. The tertiary button `SHOW MORE` should't have white stroke. It should be the same colour as the `x` button. (-> #282828)

The second thing is the size of the bold font. It is too small. It should be the same as other subheadlines. (
<img width="186" alt="image" src="https://user-images.githubusercontent.com/57008159/224632653-28471896-a8db-4be4-84af-e11924b1d36b.png">)

Hannsek commented:
![IMG_8153](https://user-images.githubusercontent.com/57008159/224666680-410aba59-8946-4ecb-b79a-8adbdcc2f730.jpg)
The header is not centered with the button.
The button has different pressed state colour than it should be. It should be _grey pressed_: #4F4F4F

Hannsek commented:
On the QR is a liitle bit close to the right top `x` button. The QR should be a little lower.
Also the header is not centered here with the `x` button.
![IMG_8154](https://user-images.githubusercontent.com/57008159/224668058-1b8f70f2-852a-470d-8aaa-b66ecec1500c.jpg)

Hannsek commented:
When I swipe between qr and address screen, it triggers some action that is sent to Suite making the Suite loads the address/Confirm on Trezor screen again and again.

matejcik commented:
![image](https://user-images.githubusercontent.com/596073/224701389-e0542965-7969-44ea-85c0-c22867004ba1.png)
the kerning for "A" in the title is obviously wrong, there's additional space to the left that should not be there

mmilata commented:
I don't think we do _any_ kerning:(

matejcik commented:
we don't do _dynamic_ kerning, yes. (maybe we'll need to start, looking at the screenshots). But the _static_ space to the left of A is wrong.

mmilata commented:
Leftmost pixel cut fix (but not `FT_LOAD_FORCE_AUTOHINT`): [ui diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3923433831/artifacts/test_ui_report/index.html) / [differing screens](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3923433831/artifacts/test_ui_report/differing_screens.html)

Hannsek commented:
It feels really broken, even other letters. I would vote for not acceptable.
<img width="203" alt="image" src="https://user-images.githubusercontent.com/57008159/224754064-40c8bfd0-f75b-438d-8448-041e9a3dcf06.png">

mmilata commented:
Sorry, in that screenshot only the BOLD font was changed, here's with all fonts fixed up + `FT_LOAD_FORCE_AUTOHINT`:

![60a9781589dac4c3da4de2f7bf86b71a0723c89772256dff43fc71fc7be45d92](https://user-images.githubusercontent.com/85857/224755599-9e2896c2-58ea-4c6f-b7cc-14ce0909796f.png)

mmilata commented:
I can also try cutting the pixel row from right side + combinations with `FT_LOAD_FORCE_AUTOHINT` but other than that I'm out of ideas. We'd have to either extend the layout engine to blend neighboring glyphs when they overlap, or implement kerning (would increase size of font data?), or both.

matejcik commented:
that last one seems good enough for now, @Hannsek ?
going forward we'll want to look into kerning, no question

Hannsek commented:
I'll try that on device and let you know, but probably yes. This looks better.

matejcik commented:
but a professional font designer should look into it for us, this is still not really good in a way that is difficult for us to tweak away: notice how the capital "S" looks shorter than the other letters

mmilata commented:
@Hannsek you can take testing firmware from this pipeline https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/804716687

Hannsek commented:
Normal SegWit account qr: 
![IMG_8155](https://user-images.githubusercontent.com/57008159/224766264-3c310f29-f86e-4a35-9def-39f7a18209a6.jpg)
It is too big.

Hannsek commented:
After a discussion, we will use `0 px` letter spacing for all fonts.

Hannsek commented:
> away: notice how the capital "S" looks shorter than the other letters

@matejcik Do you know why is that happening?

mmilata commented:
Commit 1ab09ba84a61f14cabb2fcfa55fcee5953581de8 should fix problems with title placement and corner button overlapping with QR code:

![767a77a7027a27d39f44aa9feac0d7385d08b1ebe9be43f4404f61163e8a68cb](https://user-images.githubusercontent.com/85857/224844229-06bec40e-c428-4555-9be3-70801689a6b0.png)

Hannsek commented:
![image](https://user-images.githubusercontent.com/57008159/224947705-19cc0fb9-8627-4726-b7a4-3cd93bf04f6a.png)

![image](https://user-images.githubusercontent.com/57008159/224947619-f9caad43-8d6f-4118-a30f-7b5d5ab6ad6e.png)

![image](https://user-images.githubusercontent.com/57008159/224947862-e4cfdc5d-056c-4e76-8dbe-c79f8b22a266.png)

![image](https://user-images.githubusercontent.com/57008159/224947944-c443898d-cb55-40f4-bf12-c13d042b5b27.png)


Still wrong (small) font at some places . 

QR code still valid? 
![image](https://user-images.githubusercontent.com/57008159/224946278-821acb91-d211-469e-a9c8-0b17a41a88dd.png)

I can see on device that it is ok, but not in here https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3924973481/artifacts/test_ui_report/differing_screens.html .
On device, it is quite small. Maybe, instead of making the qr code smaller, we can move it towards the bottom of the screen a bit and let the size be the same? Or maybe make it less smaller than it is now.

![image](https://user-images.githubusercontent.com/57008159/224946791-5ead9d31-f780-42a6-90cd-ba101a1978f4.png)
2-rows header not aligned.

matejcik commented:
> @matejcik Do you know why is that happening?

Pretty sure it's because the font is not designed for this combination of size + anti-aliasing.
Didn't check the pixels but my guess is the S is exactly as high as the surrounding letters, but due to the curve and anti-aliasing, it visually looks a little smaller and uneven. So a font design issue, again, very little we could do to tweak this away.

mmilata commented:
current [ui diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3934001954/artifacts/test_ui_report/index.html)/[differing screens](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3934001954/artifacts/test_ui_report/differing_screens.html)

> Still wrong (small) font at some places .

Fixed in 832fe3b41d7eb314706bce5e6c6814b2921377c6, touches a lot of altcoin screens as it affects everything that uses the `confirm_properties` layout.

![f1fd54c9b09b1293c16f7f555d38026f2716c62e4dbb234cddab9d35218507af](https://user-images.githubusercontent.com/85857/225142337-04772ddc-bb03-4506-9050-5523d78c66df.png) ![dd524363c3e3265a32970617ea4f36e8d2ef487121a85a17f54dead7509f8f7b](https://user-images.githubusercontent.com/85857/225142330-ad682880-2d10-4dd0-b591-10ca03c62587.png)


> QR code issue still valid?

Kinda but should be fixed now.

![96138586a919d5661f09d2b47d807f03cc1cdcc2c97076b4ebf58a6fe41b5778](https://user-images.githubusercontent.com/85857/225140396-3cdaa922-86a1-435c-8284-5fc202e57c59.png)

Also [832fe3b](https://github.com/trezor/trezor-firmware/pull/2876/commits/832fe3b41d7eb314706bce5e6c6814b2921377c6) adds logic to the receive screen to handle situation when xpub doesn't fit on a single screen.

![7563e399a5dcc5d550c5847a212a0b2bc1627a56adfb05c4a16504203ef55ab4](https://user-images.githubusercontent.com/85857/225141109-f0a379fd-4851-4ec3-b7a2-8810c0e90a97.png) ![84448fe22087011aab67a5a579a34c423009fa4964cc80f1dbdcdd768bdf6743](https://user-images.githubusercontent.com/85857/225141122-499ad514-bf36-46f1-874b-12e17303919e.png)

I still need to fix tests for this layout.

Hannsek commented:
![image](https://user-images.githubusercontent.com/57008159/225256946-04ef9f0b-6950-4540-aba1-2a54e4c3e4f6.png)
On that screen, we can delete the "16 tries left" text, no?

Hannsek commented:
When you want to abort recovery, there is a screen where you confirm exiting the flow. On this screen there is also **bold** font instead of demibold.

mmilata commented:
> Font still too small (bold) in coinjoin and other screens:
> When you want to abort recovery, there is a screen where you confirm exiting the flow. On this screen there is also bold font instead of demibold.

Fixed in 443e4c8307873686952a9ae47be2ce1328307459, please check [ui diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3943198037/artifacts/test_ui_report/index.html) / [differing screens](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3943198037/artifacts/test_ui_report/differing_screens.html)

> Regarding the qr, I see that the pagination marks were shifted towards the bottom. Is this intended from your side?

I moved them up a bit in dd519d936c560d200036a21c6bfa20cea1b6d579, not sure I can do much better than this:

![d1fb75040b0bd6776b8e2b79204e2fccfccbdb7233182cf530a5096cd9f17589](https://user-images.githubusercontent.com/85857/225454594-ae2f13b3-822d-4080-b5cb-8c7ed03cf64a.png)

> Together with the above, I started having problems with pressing the i/x button in top right corner. It is not responsive

Enlarged the area in 9ae05d301ee63a0f34e1b23f701372a9c9be13be to have about twice as large side than the button. I can trigger it pretty consistently now, we just need to be careful so that the touch area doesn't overlap with another button-like component.

>  On that screen, please delete the word "old".

Removed in 443e4c8307873686952a9ae47be2ce1328307459.

Hannsek commented:
> I moved them up a bit in https://github.com/trezor/trezor-firmware/commit/dd519d936c560d200036a21c6bfa20cea1b6d579, not sure I can do much better than this:

What is the blocker here?

prusnak commented:
Showing addresses and xpubs with a proportional font gives me shivers. Is there a way how to show these in non-proportional font (monospace)?

Hannsek commented:
Addresses are now shown in `Roboto Mono`, which is unproportional.
<img width="239" alt="image" src="https://user-images.githubusercontent.com/57008159/225674297-faed4020-5a0f-498b-b026-7815be7e677e.png">

We can change the xpubs font to the Roboto Mono too.

prusnak commented:
> We can change the xpubs font to the Roboto Mono too.

Please do. Also for `asset1blahblah` from screenshots whatever that is. 

Rule of thumb: anything Base58 or Bech32 encoded should be shown via monospace font.

grdddj commented on tests/device_tests/bitcoin/test_getaddress_show.py:
```diff
@@ -290,16 +290,10 @@ def input_flow():
-                    "(YOURS)" if i == xpub_num else "(COSIGNER)"
-                )
-
+            for xpub_num in range(6):
```
`pylint` complains about unused variable `xpub_num`

grdddj commented on pyproject.toml:
```diff
@@ -75,6 +75,7 @@ vulture = "^2.6"
 [tool.poetry.dev-dependencies]
 scan-build = "*"
 towncrier = "^21.9.0"
+#freetype-py = "^2.3.0"
```
Does it belong here?

mmilata replied:
I always forget the name of the package when i need to regenerate fonts. Will add it as a real dependency after #2890 is merged.

grdddj reviewed (commented):
Some comments

![image](https://user-images.githubusercontent.com/42543243/226185263-be6cf880-7ab5-4e7f-852e-816b2b4a5167.png) - there are no arrows (ellipsis) signaling previous and next page

grdddj commented:
Just tried building a debug firmware and found out it cannot fit into flash:

```
$ PYOPT=0 make build_firmware
...
/nix/store/7c8917dkn1fk7wlg8mbxdrc4i674kniz-gcc-arm-embedded-10.3.1/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld: build/firmware/firmware.elf section `.flash2' will not fit in region `FLASH2'
/nix/store/7c8917dkn1fk7wlg8mbxdrc4i674kniz-gcc-arm-embedded-10.3.1/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld: region `FLASH2' overflowed by 1544 bytes
Memory region         Used Size  Region Size  %age Used
           FLASH:        753 KB       768 KB     98.05%
          FLASH2:      919048 B       896 KB    100.17%
          CCMRAM:       31580 B        64 KB     48.19%
            SRAM:        192 KB       192 KB    100.00%
```

mmilata commented on core/embed/rust/src/ui/model_tt/component/homescreen/render.rs:
```diff
@@ -197,7 +198,7 @@ fn homescreen_line_blurred(
 
     for x in 0..HOMESCREEN_IMAGE_SIZE {
-        let c = if homescreen_dim_area(x, y) {
+        let c = if LOCKSCREEN_DIM_ALL {
```
Might come in handy should we decide that we again want to render part of the screen undimmed. The code's short, been already tested, and takes 0 flash space.

mmilata commented on core/embed/rust/src/ui/model_tt/component/address_details.rs:
```diff
@@ -7,32 +7,38 @@ use crate::{
     xpub_view: Frame<Paragraphs<Paragraph<T>>, T>,
-    xpubs: Vec<(T, T), 16>,
+    xpubs: Vec<(T, T), MAX_XPUBS>,
+    xpub_pages: Vec<u8, MAX_XPUBS>,
```
Renamed the field 22b73b8f2ad7a47e1cd168c1426cc7bf9328f7de. Having the field be usize makes the struct 48 bytes smaller, hard to say whether it's a good justification. I've at least got rid of the `as usize` casts and replaced them with `.into()` since it is defined for `u8->usize`.

mmilata commented on core/embed/rust/src/ui/model_tt/component/address_details.rs:
```diff
@@ -51,76 +57,117 @@ where
     }
 
     pub fn add_xpub(&mut self, title: T, xpub: T) -> Result<(), Error> {
+        self.xpub_pages.push(1u8).map_err(|_| Error::OutOfRange)?;
```
I wanted to maintain an invariant that the vectors always have the same length, but I suppose it's not really needed, see 22b73b8f2ad7a47e1cd168c1426cc7bf9328f7de.

mmilata commented on core/embed/rust/src/ui/model_tt/component/frame.rs:
```diff
@@ -44,6 +53,35 @@ where
         self
     }
 
+    pub fn with_button(mut self, icon: Icon, msg: CancelInfoConfirmMsg) -> Self {
```
57a3982418131712c2fb7256026a676a15138367

mmilata commented on core/embed/rust/src/ui/model_tt/component/frame.rs:
```diff
@@ -145,22 +203,18 @@ where
             title,
-            style,
-            Self::TEXT_OFFSET,
+            Font::BOLD,
```
oof yea, 57a3982418131712c2fb7256026a676a15138367

mmilata commented:
Freed up around 4KiB with [eade24c](https://github.com/trezor/trezor-firmware/pull/2876/commits/eade24c5cb4481e60a9c5a0be2f0f300f0da7bc6) by adding a cargo flag for the `Component::bounds()` method, which is only very rarely used.

xpubs use monospace font now:
![33b84fa3c1110c257503d99c5992908ff4e4da13ecd4db9247199255eff43656](https://user-images.githubusercontent.com/85857/226764172-0ff74c10-a62d-4ad5-b587-d3f826721d25.png)

Hannsek commented:
Text in buttons are not vertically centered.
![image](https://user-images.githubusercontent.com/57008159/226829586-2714a15c-d5b0-4fd0-837e-6513f966668f.png)
![image](https://user-images.githubusercontent.com/57008159/226829802-a51bb20a-d12c-4d11-b009-7150b0046dc7.png)
![image](https://user-images.githubusercontent.com/57008159/226829926-ff606d15-e031-4763-946a-1249ac803246.png)
But i don't see it as a blocker. If it is not an easy fix, I will create new issue.

mmilata commented on core/embed/rust/src/ui/model_tt/component/homescreen/render.rs:
```diff
@@ -197,7 +198,7 @@ fn homescreen_line_blurred(
+        let c = if LOCKSCREEN_DIM_ALL {
             let x = x as usize;
 
             let coef = (65536_f32 * LOCKSCREEN_DIM) as u32;
```
But how to call it?

TychoVrahe replied:
this is basically number one in Q0.16 format (or more precisely serves as reference to that when calculating numbers in said format). 

so either we can name in like that, or maybe we could introduce a macro called i.e. Q0_16 where this constant would be used for calculating the values

mmilata commented:
Fixed vertical centering in [8b51882](https://github.com/trezor/trezor-firmware/pull/2876/commits/8b51882804a63303ad32bfd5152461d7e9a78a26) and changed paging a bit in [298e0e4](https://github.com/trezor/trezor-firmware/pull/2876/commits/298e0e4561d479725f031c9260a3fe2aa77c1fa7).

![290744139aa26b0e6b211fd909cc8e70a82a8c63993f22fbf664e207e8eeafea](https://user-images.githubusercontent.com/85857/227018404-9e64cc95-f52b-4848-bea1-55692f3509bc.png) ![bb54d47315ba8dee8acf0b03ee16f8c10426f0c15dff691c4d9fc5faf0b6b2ed](https://user-images.githubusercontent.com/85857/227018422-88208977-5a56-4514-9521-b55582267497.png) ![Screenshot 2023-03-22 at 20-36-47 TT-device_tests-bitcoin-test_signmessage py test_signmessage segwit-native long message](https://user-images.githubusercontent.com/85857/227018463-484ca287-5628-4117-8b15-88c61034676c.png)

mmilata commented:
Rebased, [master diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3984714030/artifacts/master_diff/index.html).

grdddj commented:
> LGTM, I will revert the `master` `fixtures.json` changes, and will ping you after they are done

Changes done in https://github.com/trezor/trezor-firmware/pull/2903 - hopefully soon to be merged

bosomt commented:
**QA OK**

if needed we create separate issue, so far i was not able to find any issue in latest UI 
 - Device: model T 2.6.0 Universal (revision 6e0c2578f8932c94413b2d1564725dc687eceaf7)
